Add read-only role page to documents #5783 (#6806)
opensearch-project/documentation-website: _security/access-control/read-only-role.md

---
layout: default
title: Read only roles
parent: Access control
nav_order: 150
---

# Read only roles

> As with any roles in OpenSearch, read only roles can be configured using three methods: `yml` configuration files, API and OpenSearch Dashboards. The most user friendly approach for anyone getting familiar with Roles and role mappings is to use OpenSearch Dashboards, as it is easy to navigate the creation of roles and assignment to users. The basic steps of creating roles/mappings/users are outlines here (https://opensearch.org/docs/latest/security/access-control/users-roles/)

Reviewer: ```suggestion
As with any roles in OpenSearch, read only roles can be configured using three methods: `yml` configuration files, API and OpenSearch Dashboards. The most user friendly approach for anyone getting familiar with Roles and role mappings is to use OpenSearch Dashboards, as it is easy to navigate the creation of roles and assignment to users. The basic steps of creating roles/mappings/users are outlined here (https://opensearch.org/docs/latest/security/access-control/users-roles/)
```

## Basic read only role

> If you are looking to create a role to access OpenSearch Dashboards, view existing dashboards/visualizations and query different indexes, assuming you want the user to have read access to all indexes and tenants, you can add cluster permission group of `cluster_composite_ops_ro`, `read` access to all indexes (`*`) and `read` access to all tenants

Reviewer: Not sure if we should use "indices." Ask @Naarcha-AWS

> If you are looking to create a role to access OpenSearch Dashboards, view existing dashboards/visualizations and query different indexes, assuming you want the user to have read access to all indexes and tenants, you can add cluster permission group of `cluster_composite_ops_ro`, `read` access to all indexes (`*`) and `read` access to all tenants

Reviewer: I would consider explaining each of these. Without knowing the background it is not super obvious what `cluster_composite_ops_ro` is for. Likewise it may not be obvious why you restate the permissions at different levels so you may want to clarify this.

Author: @scrawfor99 good point, I've expanded on this, let me know if below is better:
## Basic read only role

If you are looking to create a role to access OpenSearch Dashboards, view existing dashboards/visualizations and query different indexes, assuming you want the user to have read access to all indexes and tenants, you can use the following permissions.
### Cluster permission
Cluster permission is needed to access cluster wide resources like visualizations and dashboards. For read only user you add already created `cluster_composite_ops_ro` permission.
### Indexes permission
In order to view visualizations, user needs to have access to the index that was used to create it, In this can you can add `read` permission to all (`*`) indexes. 
### Tenant permissions
If you are using tenants to separate work area between different teams/projects, visualizations and dashboards are saved in specific tenant. To be able to view these the user needs `read` access to the correct tenant. To keep things more straightforward we are giving user access to all (`*`) tenants.

![creating role]({{site.url}}{{site.baseurl}}/images/role_creation_read_only.png)

> Once the role is created, you can directly map the role to a user by going to "Mapped users" tab in roles, select "Map users" and select the user to map to this role. 

Reviewer: @Naarcha-AWS do you know if this is the correct way to mention buttons?

![mapping users]({{site.url}}{{site.baseurl}}/images/mapping-users.png)

## OpenSearch Dashboards readonly_mode

> OpenSearch Dashboards comes with functionality to only give access to `Dashboards` UI by using readonly_mode.

Reviewer: Not sure on stylization of "readonly_mode;" does it need to be in quotes?

Author: I've added code to readonly_mode

> OpenSearch Dashboards comes with functionality to only give access to `Dashboards` UI by using readonly_mode.

Reviewer: This line reads a little confusing in my opinion maybe rephrase? It is not clear _what_ access is given until you see the readonly_mode term.

Author: @scrawfor99 I've reworded this now to:
OpenSearch Dashboards `readonly_mode` functionality is used to give user access to only `Dashboards` UI, removing the rest of the UI elements from the user.

You need to configure this in `opensearch_dashboards.yml` file by adding following line:
`opensearch_security.readonly_mode.roles: [new_role]`

If the role mapped to user has additional privileges or user is mapped to other roles, giving them write access to indexes, this access will not be allowed using OpenSearch Dashboard. Direct data access to OpenSearch using curl or API is still allowed, as OpenSearch Dashboards is not involved in this communication. 

> If the user is mapped to this `readonly_mode` role, all other elements of the UI will be removed, except for `Dashboards`. See following comparison, on the left is user mapped to role which is configured as readonly_mode, on the right is the standard view.

Reviewer: You write readonly_mode as code here but not above.

Author: @scrawfor99 thanks for that Stephen, I've update to code now

> If the user is mapped to this `readonly_mode` role, all other elements of the UI will be removed, except for `Dashboards`. See following comparison, on the left is user mapped to role which is configured as readonly_mode, on the right is the standard view.

Reviewer: ```suggestion
If the user is mapped to this `readonly_mode` role, all other elements of the UI will be removed, except for `Dashboards`. In the following comparison, the left view shows the screen from the perspective of a user mapped to a role which is configured as readonly_mode. On the right, the user is given the standard view.
```

---

# Read only roles

As with any roles in OpenSearch, read only roles can be configured using three methods: `yml` configuration files, API and OpenSearch Dashboards. The most user friendly approach for anyone getting familiar with Roles and role mappings is to use OpenSearch Dashboards, as it is easy to navigate the creation of roles and assignment to users. The basic steps of creating roles/mappings/users are outlined here (https://opensearch.org/docs/latest/security/access-control/users-roles/)

## Basic read only role

If you are looking to create a role to access OpenSearch Dashboards, view existing dashboards/visualizations and query different indexes, assuming you want the user to have read access to all indexes and tenants, you can use the following permissions.
### Cluster permission
Cluster permission is needed to access cluster wide resources like visualizations and dashboards. For read only user you add already created `cluster_composite_ops_ro` permission.
### Indexes permission
In order to view visualizations, user needs to have access to the index that was used to create it, In this can you can add `read` permission to all (`*`) indexes. 
### Tenant permissions

> If you are using tenants to separate work area between different teams/projects, visualizations and dashboards are saved in specific tenant. To be able to view these the user needs `read` access to the correct tenant. To keep things more straightforward we are giving user access to all (`*`) tenants.

Reviewer 2: ```suggestion

If you use tenants to split work between different teams or projects, use the all (`*`) option followed by the **Read only** option.
```

![creating role]({{site.url}}{{site.baseurl}}/images/role_creation_read_only.png)

> Once the role is created, you can directly map the role to a user by going to **Mapped users** tab in roles, select **Map users** and select the user to map to this role. 

Reviewer 2: ```suggestion
After all permission types are set and the role is created, you can directly map the role to a user by going to **Mapped users** tab in roles. Select **Map users** and select the user to map to this role. 
```

![mapping users]({{site.url}}{{site.baseurl}}/images/mapping-users.png)

## OpenSearch Dashboards readonly_mode

> OpenSearch Dashboards `readonly_mode` functionality is used to give user access to only `Dashboards` UI, removing the rest of the UI elements from the user.

Reviewer 2: Can we give an example of the other UI elements?

```suggestion
OpenSearch Dashboards `readonly_mode` functionality is used to give a user access to only the `Dashboards` UI, removing the rest of the UI elements from the user.
```

> You need to configure this in `opensearch_dashboards.yml` file by adding following line:

Reviewer 2: ```suggestion
To add the functionality, add the following line `opensearch_dashboards.yml` file:
```

Reviewer 2: We need to wrap this in the ``` marks.

`opensearch_security.readonly_mode.roles: [new_role]`

> If the role mapped to user has additional privileges or user is mapped to other roles, giving them write access to indexes, this access will not be allowed using OpenSearch Dashboard. Direct data access to OpenSearch using curl or API is still allowed, as OpenSearch Dashboards is not involved in this communication. 

Reviewer 2: ```suggestion
If the role mapped to the user has additional privileges or the user is mapped to other roles, giving them write access to indexes, this access will not be allowed using OpenSearch Dashboard. Direct data access to OpenSearch using curl or API is still allowed, as OpenSearch Dashboards is not involved in this communication. 
```

> If the user is mapped to this `readonly_mode` role, all other elements of the UI will be removed, except for `Dashboards`. In the following comparison, the left view shows the screen from the perspective of a user mapped to a role which is configured as readonly_mode. On the right, the user is given the standard view.

Reviewer 2: ```suggestion
If the user is mapped to the `readonly_mode` role, all other elements of the UI will be removed, except for `Dashboards`. In the following comparison, the left view shows the screen from the perspective of a user mapped to a role in`readonly_mode`. On the right, the user is given a standard view.
```

![compare read only mode]({{site.url}}{{site.baseurl}}/images/compare_read_only_mode.png)

> Don't forget that mapping the user to only readonly_mode role, that doesn't give permissions to view relevant indexes, will not allow the user to view the existing dashboards, as read access is still required to view data behind the dashboards.

Reviewer 2: ```suggestion
Mapping the user to only `readonly_mode` role does not give permissions to view relevant indexes or allow the user to view the existing dashboards. Read access to indexes and dashboards require separate permissions.
```

{: .note }

If the user is also mapped to any roles listed under `plugins.security.restapi.roles_enabled` in `opensearch.yml`, for example `all_access` or `security_rest_api_access`, then `readonly_mode` is ignored, giving the user access to standard UI elements.

## Additional permissions

> If you need access to additional permissions while using read_only role, for example for alerting and anomaly detection modules, check out the existing roles, for example "alerting_read_access" and "anomaly_read_access"

Reviewer 2: ```suggestion
If you need access to additional permissions while using the `read_only` role,  such as alerting and anomaly detection modules, check out the existing roles, such as `alerting_read_access` and `anomaly_read_access`.
```
